### PR TITLE
Temporarily adding CloudShellCredentials to our project, with a fix.

### PR DESCRIPTION
**Pull request by sduskis**
Currently, the quickstart hbase shell doesn't work in the cloud shell.  There's a bug in in the google java auth, for which I'll submit a PR.  In the mean time, this change in this repo allows us to use the hbase shell in the Cloud Shell

### CODE REVIEW

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java [hunk 1]
```diff
@@ -90,24 +91,30 @@
 
     final IOException exception;
     final String header;
-    final long staleTimeMs;
-    final long expiresTimeMs;
+    final Long staleTimeMs;
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java]: can you document that when staleTimeMs is null, the cache will be stale or expired?

**Comment by kevinsi4508** [comment]
newly added CloudShellCredentials is not used anywhere.

**Comment by sduskis** [comment]
I added comments to help with the issues you mentioned.  I hope that the CloudShellCredentials comment addresses your question well enough.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java [hunk 2]
```diff
@@ -86,28 +89,44 @@
      * After the token is "expired," the interceptor blocks gRPC calls. The Expired state indicates
      * that the interceptor needs to do a synchronous refresh.
      */
-    public static final int TOKEN_EXPIRES_MS = 45 * 1000;
+    public static final int TOKEN_EXPIRES_MS =5 * 1000;
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java]: leave a space after "="

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java [hunk 3]
```diff
@@ -86,28 +89,44 @@
...
 
     final IOException exception;
     final String header;
-    final long staleTimeMs;
-    final long expiresTimeMs;
+    
+    /**
+     * Defines the amount of time in ms when the header is considered "stale" and should be
+     * refreshed in the near future.  A null value means that the header does not become stale.
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java]: remove an extra white space "future. A"
"{@code null} value means ..." ?

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java [hunk 4]
```diff
@@ -86,28 +89,44 @@
...
-      Preconditions.checkState(staleTimeMs < expiresTimeMs);
+      Date expirationTime = token.getExpirationTime();
+      if (expirationTime != null) {
+        long tokenExpiresTime = expirationTime.getTime();
+        this.staleTimeMs = tokenExpiresTime - TOKEN_STALENESS_MS;
+        // Block until refresh at this point.
+        this.expiresTimeMs = tokenExpiresTime - TOKEN_EXPIRES_MS;
+        Preconditions.checkState(staleTimeMs < expiresTimeMs);
+      } else {
+        this.staleTimeMs = null;
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java]: add a test case to test when these are set to null, header doesn't become stale or expired?

Comment (sduskis) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java]: Good idea.  Done.

**Comment by sduskis** [comment]
I added a test, and updated the comments.  PTAL at your convenience.

File: bigtable-client-core/src/main/java/com/google/auth/oauth2/CloudShellCredentials.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+/**
+ * OAuth2 credentials representing the built-in service account for Google Cloud Shell. See
+ * https://cloud.google.com/cloud-shell/docs/ for more information.
+ * <p>
+ * NOTE: this is a copy of
+ * https://github.com/google/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/
+ * oauth2/CloudShellCredentials.java with a bug fix. This class gets called from
+ * {@link CredentialOptions#defaultCredentials()} if cloud bigtable is invoked from the Google Cloud
+ * Shell.  This copy should be removed once the google-auth-library-java has the fix.
+ * <p>
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/auth/oauth2/CloudShellCredentials.java]: Can you indicate what is the fix here?

File: bigtable-client-core/src/main/java/com/google/auth/oauth2/CloudShellCredentials.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ */
+public class CloudShellCredentials extends GoogleCredentials {
+
+  private final static int ACCESS_TOKEN_INDEX = 2;
+  private final static int READ_TIMEOUT_MS = 5000;
+
+  /**
+   * The Cloud Shell back authorization channel uses serialized
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/auth/oauth2/CloudShellCredentials.java]: The Cloud Shell backed?

File: bigtable-client-core/src/main/java/com/google/auth/oauth2/CloudShellCredentials.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+   * Javascript Protobufers, preceeded by the message lengeth and a
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/auth/oauth2/CloudShellCredentials.java]: preceded by the message length

**Comment by sduskis** [comment]
I added a link to the PR in the other repo.  I think it's a clearer indication of the fix than any text I could provide.

**Comment by kevinsi4508** [comment]
LGTM
